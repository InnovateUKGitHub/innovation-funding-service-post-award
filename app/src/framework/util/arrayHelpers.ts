/**
 * returns array with null or undefined elements removed
 */
export function filterEmpty<T>(array: (T | null | undefined)[]): T[] {
  return array.filter(x => x !== null && x !== undefined) as T[];
}

/**
 * Returns an array flattened one level
 */
export function flatten<T>(array: T[][]): T[] {
  return array.reduce<T[]>((a, b) => a.concat(b), []);
}

export const hasNoDuplicates = <T>(array: T[]) => array.every((e, i) => array.indexOf(e) === i);

export const groupBy = <T, K>(list: T[], getKey: (entry: T) => K) => {
  const map = new Map<K, T[]>();
  list.forEach(entry => {
    const key = getKey(entry);
    const values = map.get(key) || [];
    map.set(key, [...values, entry]);
  });
  return map;
};

export const getArrayFromPeriod = <T extends { periodId: number }>(
  originalArray: T[],
  currentPeriod: number,
  lastPeriodId?: number,
) => {
  if (!originalArray.length) return originalArray;

  return originalArray.filter(item => {
    const notPreviousPeriod = item.periodId >= currentPeriod;

    if (!lastPeriodId) return notPreviousPeriod;

    return notPreviousPeriod && item.periodId <= lastPeriodId;
  });
};

export const getArrayExcludingPeriods = <
  T extends {
    periodId: number;
  },
>(
  originalArray: T[],
  excludePeriods: Set<number>,
): T[] => {
  if (!originalArray.length) return originalArray;

  return originalArray.filter(item => !excludePeriods.has(item.periodId));
};
