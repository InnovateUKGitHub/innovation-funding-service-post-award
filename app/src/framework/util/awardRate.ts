import { AwardRateOverrideType } from "@framework/constants/awardRateOverride";
import { ClaimOverrideRateDto } from "@framework/dtos/claimOverrideRate";
import { PartnerDto } from "@framework/dtos/partnerDto";

interface GetAwardRateProps {
  partner: Pick<PartnerDto, "awardRate">;
  claimOverrides: ClaimOverrideRateDto;
  periodId: PeriodId;
  costCategoryId: CostCategoryId;
}

const getAwardRate = ({ partner, claimOverrides, periodId, costCategoryId }: GetAwardRateProps) => {
  const baseAwardRate = partner.awardRate ?? 0;

  if (claimOverrides.type === AwardRateOverrideType.BY_COST_CATEGORY) {
    const override = claimOverrides.overrides.find(x => x.costCategoryId === costCategoryId);
    if (override) return override.amount;
    return baseAwardRate;
  }

  if (claimOverrides.type === AwardRateOverrideType.BY_PERIOD) {
    const override = claimOverrides.overrides.find(x => x.period === periodId);
    if (override) return override.amount;
    return baseAwardRate;
  }

  return baseAwardRate;
};

export { getAwardRate };
