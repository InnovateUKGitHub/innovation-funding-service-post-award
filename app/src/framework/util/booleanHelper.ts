export const isBoolean = (value: boolean | null | undefined): value is boolean => value === true || value === false;
