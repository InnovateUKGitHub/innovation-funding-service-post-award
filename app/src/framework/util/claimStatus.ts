import { PartnerClaimStatus } from "@framework/constants";

/**
 * converts salesforce labels to claim status enum
 */
export function getClaimStatus(salesforceStatus: string): PartnerClaimStatus {
  switch (salesforceStatus) {
    case "No Claims Due":
      return PartnerClaimStatus.NoClaimsDue;
    case "Claim Due":
      return PartnerClaimStatus.ClaimDue;
    case "Claims Overdue":
      return PartnerClaimStatus.ClaimsOverdue;
    case "Claim Queried":
      return PartnerClaimStatus.ClaimQueried;
    case "Claim Submitted":
      return PartnerClaimStatus.ClaimSubmitted;
    case "Awaiting IAR":
      return PartnerClaimStatus.IARRequired;
    default:
      return PartnerClaimStatus.Unknown;
  }
}
