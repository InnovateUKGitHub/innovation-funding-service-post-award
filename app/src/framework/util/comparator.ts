import { DateTime } from "luxon";

export type IComparer<T> = (x: T, y: T) => number;

export const numberComparator: IComparer<number> = (a, b) => a - b;

export const dateComparator: IComparer<Date> = (a, b) => numberComparator(a.getTime(), b.getTime());

export const stringComparator: IComparer<string> = (a, b) => {
  if(a && b) return a.localeCompare(b);
  if(a) return -1;
  if(b) return 1;
  return 0;
};

export const dayComparator: IComparer<Date> = (a, b) => {
  return DateTime.fromJSDate(a).startOf("day").toMillis() - DateTime.fromJSDate(b).startOf("day").toMillis();
};
