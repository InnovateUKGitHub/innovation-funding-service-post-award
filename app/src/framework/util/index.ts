export * from "./booleanHelper";
export * from "./comparator";
export * from "./dtoHelpers";
export * from "./dateHelpers";
export * from "./numberHelper";
export * from "./arrayHelpers";
export * from "./windowHelpers";
export * from "./clock";
export * from "./files";
