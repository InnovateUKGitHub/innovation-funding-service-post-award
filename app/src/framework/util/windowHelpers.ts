import { useEffect } from "react";

export const scrollToTheTopSmoothly = () =>
  requestAnimationFrame(() => window.scrollTo({ top: 0, behavior: "smooth" }));
export const scrollToTheTopInstantly = () => requestAnimationFrame(() => window.scrollTo(0, 0));

/**
 * Find a HTML element and scroll the page to the closest parent instance of FIELDSET.
 *
 * @param tag The HTML `name` or `id` to scroll to
 */
export const scrollToTheParentFieldsetSmoothly = (tag: string) => {
  /**
   * Find the closest HTML parent, starting from a child and climbing up to the parent.
   *
   * @param child The HTML element to start searching from
   * @param element The target HTML element, in uppercase characters
   * @returns The found HTML element, or null if not found.
   */
  const getParentElement = (child: HTMLElement, element: string): HTMLElement | null => {
    if (child.tagName === element) return child;
    if (child.parentElement) return getParentElement(child.parentElement, element);
    return null;
  };

  // Wait for the next animation frame before scrolling
  requestAnimationFrame(() => {
    // Find the first instance of the NAME or ID
    const [nameTag] = [...document.getElementsByName(tag), document.getElementById(tag)];

    // If it exists, scroll to it.
    if (nameTag) {
      const parentFieldset = getParentElement(nameTag, "FIELDSET");

      // If the parent fieldset exists, scroll to that.
      // Otherwise, just scroll to the tag.
      if (parentFieldset) {
        parentFieldset.scrollIntoView({ behavior: "smooth" });
      } else {
        nameTag.scrollIntoView({ behavior: "smooth" });
      }
    }
  });
};

export const useScrollToTopSmoothly = (dependencies: unknown[]) => {
  useEffect(() => {
    scrollToTheTopSmoothly();
  }, dependencies);
};
