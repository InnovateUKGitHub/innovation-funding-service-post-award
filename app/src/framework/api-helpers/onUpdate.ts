import { isApiError } from "@framework/util/errorHelpers";
import { useApiErrorContext } from "@ui/context/api-error";
import { noop } from "@ui/helpers/noop";
import { useState } from "react";
import { Logger } from "@shared/developmentLogger";
import { scrollToTheTopSmoothly } from "@framework/util/windowHelpers";

/**
 * ### useOnUpdate
 * Takes an update api call and an optional onSuccess and onError callback.
 *
 * It returns an object with the onUpdate function and apiError response.
 * apiError is also initialised with any serverRenderedApiError, mainly
 * to support js disabled
 *
 * @returns an object with `onUpdate` function and `apiError` response
 */

export const useOnUpdate = <TFormValues, TResponse, TContext = undefined>({
  req,
  onSuccess = noop,
  onError = noop,
}: {
  req: (data: TFormValues, context?: TContext) => Promise<TResponse>;
  onSuccess?: (data: TFormValues, res: TResponse, context?: TContext) => void;
  onError?: (e: unknown) => void;
}) => {
  const serverRenderedApiError = useApiErrorContext();
  const [apiError, setApiError] = useState(serverRenderedApiError);
  const [isFetching, setIsFetching] = useState(false);

  const onUpdate = async ({ data, context }: { data: TFormValues; context?: TContext }) => {
    try {
      setIsFetching(true);
      setApiError(null);

      const res = await req(data, context);

      setIsFetching(false);
      onSuccess(data, res, context);
    } catch (e: unknown) {
      const logger = new Logger("onUpdate");

      if (isApiError(e)) {
        setApiError(e);
        scrollToTheTopSmoothly();
      }

      setIsFetching(false);

      logger.error("request error", e);
      onError(e);
    }
  };

  return { onUpdate, apiError, isFetching };
};
