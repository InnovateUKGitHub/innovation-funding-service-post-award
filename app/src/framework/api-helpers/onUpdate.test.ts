import { renderHook, act } from "@testing-library/react";
import { useOnUpdate } from "./onUpdate";
import * as windowHelpers from "@framework/util/windowHelpers";

jest.spyOn(windowHelpers, "scrollToTheTopSmoothly").mockImplementation();

describe("useOnUpdate", () => {
  const onSuccess = jest.fn();
  const onError = jest.fn();
  global.window.scrollTo = jest.fn();

  const renderSuccessHook = () =>
    renderHook(() => useOnUpdate({ req: data => Promise.resolve(data), onSuccess, onError }));

  const renderFailHook = () =>
    renderHook(() =>
      useOnUpdate({
        req: () => Promise.reject({ code: 1, message: "Leo forgot his mama noodles" }),
        onSuccess,
        onError,
      }),
    );

  beforeEach(jest.clearAllMocks);
  it("should return an object with on `onUpdate` and an `apiError` property", () => {
    const { result } = renderSuccessHook();

    expect(result.current).toMatchSnapshot();
  });

  it("should call the success method and no errors if the promise resolves", async () => {
    const { result } = renderSuccessHook();

    await act(() => result.current.onUpdate({ data: { postcode: "SN1 2AP" } }));

    expect(result.current.apiError).toBe(null);
    expect(onSuccess).toHaveBeenCalledWith({ postcode: "SN1 2AP" }, { postcode: "SN1 2AP" }, undefined);
    expect(onError).not.toHaveBeenCalled();
  });

  it("should call the error method and show the apiError if the promise rejects", async () => {
    const { result } = renderFailHook();
    await act(() => result.current.onUpdate({ data: { postcode: "SN1 2AP" } }));

    expect(result.current.apiError).toEqual({ code: 1, message: "Leo forgot his mama noodles" });
    expect(onSuccess).not.toHaveBeenCalled();
    expect(onError).toHaveBeenCalled();
  });

  it("should pass in an input ctx to the success callback", async () => {
    const { result } = renderSuccessHook();
    await act(() => result.current.onUpdate({ data: { poland: "Poland" }, context: { bulgoggi: "Burger" } }));

    expect(result.current.apiError).toBe(null);
    expect(onSuccess).toHaveBeenCalledWith({ poland: "Poland" }, { poland: "Poland" }, { bulgoggi: "Burger" });
    expect(onError).not.toHaveBeenCalled();
  });
});
