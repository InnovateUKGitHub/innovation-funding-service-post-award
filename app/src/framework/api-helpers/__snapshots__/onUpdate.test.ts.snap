// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useOnUpdate should return an object with on \`onUpdate\` and an \`apiError\` property 1`] = `
{
  "apiError": null,
  "isFetching": false,
  "onUpdate": [Function],
}
`;
